.github/workflows/main.yml

```diff
@@ -88,37 +83,113 @@ jobs:
 
     - name: Build
       run: |
-        if [ "${{matrix.mode}}" == "test" ]; then
            make debug
-        else
-           make release
-        fi
 
     - name: Test
-      if: matrix.mode == 'test'
       run: |
         make test/unittest-d
 
-    - name: Regression
-      timeout-minutes: 180
-      if: matrix.mode == 'regression'
+  linux-regression:
+    name: "Regression ${{ matrix.compiler }} ${{matrix.shard}} (Linux)"
+
+    runs-on: ubuntu-latest
+    defaults:
+      run:
+        shell: bash
+    strategy:
+      fail-fast: false
+      matrix:
+        num_shards: [3]
+        shard: [0, 1, 2]
+        compiler:
+        - gcc
+        - clang
+
+    env:
+      COMPILER: ${{ matrix.compiler }}
+
+    steps:
+
+    - name: Cancel previous
+      uses: styfle/cancel-workflow-action@0.11.0
+      with:
+        access_token: ${{ github.token }}
+
+    - name: Install dependencies
       run: |
-        make test/regression
-        git status
+        sudo apt-get update -qq
+        sudo apt install -y \
+          g++-9 clang \
+          default-jre \
+          cmake ninja-build \
+          build-essential \
+          google-perftools \
+          libgoogle-perftools-dev \
+          python3 python3-orderedmultidict python3-psutil python3-dev \
+          uuid-dev \
+          lcov
+
+    - name: Checkout code
+      uses: actions/checkout@v3
+      with:
+        submodules: recursive
+        fetch-depth: 0
+
+    - name: Use ccache
+      uses: hendrikmuhs/ccache-action@v1.2
+      with:
+        key: linux-regression-${{ matrix.compiler }}
+
+    - name: Configure compiler
+      run: |
+        if [ "${{matrix.compiler}}" == "clang" ]; then
+          echo 'CC=clang' >> $GITHUB_ENV
+          echo 'CXX=clang++' >> $GITHUB_ENV
+        else
+          echo 'CC=gcc-9' >> $GITHUB_ENV
+          echo 'CXX=g++-9' >> $GITHUB_ENV
+        fi
+
+    - name: Configure shell
+      run: |
+        echo 'PATH=/usr/lib/ccache:'"$PATH" >> $GITHUB_ENV
+        echo 'PREFIX=${GITHUB_WORKSPACE}/install' >> $GITHUB_ENV
+        echo "ADDITIONAL_CMAKE_OPTIONS=-DPython3_ROOT_DIR=${pythonLocation} -DMY_CXX_WARNING_FLAGS='-W -Wall -Wextra -Wno-unused-parameter -Werror -UNDEBUG'" >> $GITHUB_ENV
+        echo "CMAKE_GENERATOR=Ninja" >> $GITHUB_ENV
+
+    - name: Show shell configuration
+      run: |
+        env
+        which cmake && cmake --version
+        which java && java -version
+        which python && python --version
+        which ninja && ninja --version
+        which $CC && $CC --version
+        which $CXX && $CXX --version
+
+    - name: Build
```

Comment:
curious if you looked at reusing something like https://github.com/marketplace/actions/run-cmake ?